File: java/src/main/java/noie/linmimeng/noiedigitalsystem/api/adapter/AssetAdapter.java
package noie.linmimeng.noiedigitalsystem.api.adapter;

import noie.linmimeng.noiedigitalsystem.api.asset.AssetId;
import noie.linmimeng.noiedigitalsystem.api.asset.AssetScope;
import noie.linmimeng.noiedigitalsystem.api.asset.NdsAsset;
import java.util.Map;
import java.util.HashMap;

/**
 * [Index] NDS-JAVA-ASSET-000
 * [Semantic] AssetId / NdsAsset Domain ↔ Proto conversion.
 *
 * @since 2.0.0
 */
public final class AssetAdapter {
    
    private AssetAdapter() {
        // [Index] NDS-JAVA-ASSET-001 [Constraint] Utility class; instantiation is prohibited.
    }
    
    // ========================================================================
    // [Index] NDS-JAVA-ASSET-010 [Semantic] AssetId conversion.
    // ========================================================================
    
    /**
     * @param domain domain AssetId (nullable)
     * @return Proto AssetId; null if input is null
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetId toProto(AssetId domain) {
        if (domain == null) {
            return null;
        }
        
        return noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetId.newBuilder()
            .setName(domain.name())
            .setScope(toProtoScope(domain.scope()))
            .build();
    }
    
    /**
     * @param proto Proto AssetId (nullable)
     * @return domain AssetId; null if proto is null or name is empty
     */
    public static AssetId fromProto(noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetId proto) {
        if (proto == null || proto.getName().isEmpty()) {
            return null;
        }
        
        AssetScope scope = fromProtoScope(proto.getScope());
        return AssetId.of(scope, proto.getName());
    }
    
    // ========================================================================
    // [Index] NDS-JAVA-ASSET-020 [Semantic] AssetScope conversion.
    // ========================================================================
    
    /**
     * @param scope domain AssetScope (nullable; treated as UNKNOWN if null)
     * @return Proto AssetScope
     */
    public static noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope toProtoScope(AssetScope scope) {
        if (scope == null) {
            return noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope.ASSET_SCOPE_UNSPECIFIED;
        }
        
        return switch (scope) {
            case PLAYER -> noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope.ASSET_SCOPE_PLAYER;
            case SERVER -> noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope.ASSET_SCOPE_SERVER;
            case GLOBAL -> noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope.ASSET_SCOPE_GLOBAL;
            case UNKNOWN -> noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope.ASSET_SCOPE_UNSPECIFIED;
        };
    }
    
    /**
     * @param protoScope Proto AssetScope (nullable; treated as UNKNOWN if null)
     * @return domain AssetScope
     */
    public static AssetScope fromProtoScope(noie.linmimeng.noiedigitalsystem.api.proto.asset.AssetScope protoScope) {
        if (protoScope == null) {
            return AssetScope.UNKNOWN;
        }
        
        return switch (protoScope) {
            case ASSET_SCOPE_PLAYER -> AssetScope.PLAYER;
            case ASSET_SCOPE_SERVER -> AssetScope.SERVER;
            case ASSET_SCOPE_GLOBAL -> AssetScope.GLOBAL;
            case ASSET_SCOPE_UNSPECIFIED, UNRECOGNIZED -> AssetScope.UNKNOWN;
        };
    }
    
    // ========================================================================
    // [Index] NDS-JAVA-ASSET-900 [Semantic] NdsAsset conversion.
    // [Trace] Reserved for future expansion.
    // ========================================================================
    
    // [Index] NDS-JAVA-ASSET-901 [TODO] Implement full NdsAsset <-> proto conversion when required.
    // public static noie.linmimeng.noiedigitalsystem.api.proto.asset.NdsAsset toProto(NdsAsset domain)
    // public static NdsAsset fromProto(noie.linmimeng.noiedigitalsystem.api.proto.asset.NdsAsset proto)
}

File: csharp/src/Noie.Nds.Api/Adapter/AssetAdapter.cs
using Noie.Nds.Api.Asset;

namespace Noie.Nds.Api.Adapter;

/// <summary>
/// [Index: NDS-CSHARP-ASSETADAPTER-000] Asset adapter (IAssetId ↔ proto-compatible shapes).
/// 
/// <para>[Semantic] Conversion utilities for asset scopes and canonical ID formats.</para>
/// </summary>
public static class AssetAdapter
{
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-010] Convert domain AssetScope to a proto string token.
    /// </summary>
    public static string ToProtoString(AssetScope scope)
    {
        return scope.ToProtoString();
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-011] Convert a proto string token to domain AssetScope.
    /// </summary>
    public static AssetScope FromProtoString(string? protoScope)
    {
        return AssetScopeExtensions.FromString(protoScope);
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-020] Serialize a domain asset ID to canonical string form.
    /// </summary>
    /// <param name="assetId">Domain asset ID.</param>
    /// <returns>Canonical full ID string.</returns>
    public static string Serialize(IAssetId assetId)
    {
        return assetId.FullId;
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-021] Deserialize canonical string form into a domain asset ID.
    /// </summary>
    /// <param name="fullId">Full ID string.</param>
    /// <returns>Domain asset ID.</returns>
    public static IAssetId Deserialize(string fullId)
    {
        return Asset.AssetId.FromString(fullId);
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-030] Create a proto-compatible asset ID tuple.
    /// </summary>
    public static (string Name, string Scope) ToProtoData(IAssetId assetId)
    {
        return (assetId.Name, assetId.Scope.ToProtoString());
    }
    
    /// <summary>
    /// [Index: NDS-CSHARP-ASSETADAPTER-031] Create a domain asset ID from proto-compatible data.
    /// </summary>
    public static IAssetId FromProtoData(string name, string scope)
    {
        var assetScope = AssetScopeExtensions.FromString(scope);
        return Asset.AssetId.Of(assetScope, name);
    }
}
